kubectl-cloudflow/version/version.go

```diff
@@ -30,40 +32,70 @@ const ProtocolVersionKey = "protocol-version"
 // ProtocolVersionConfigMapName is the name of the configmap that contains the protocol-version
 const ProtocolVersionConfigMapName = "cloudflow-protocol-version"
 
-// FailOnProtocolVersionMismatch fails and exits if the protocol version of kubectl-cloudflow does not match with the cloudflow operator protocol version.
-func FailOnProtocolVersionMismatch() {
+// GetProtocolVersionConfigMap Get the protocol version config map set by the operator
+func GetProtocolVersionConfigMap() (*corev1.ConfigMap, error) {
 	k8sClient, k8sErr := k8s.GetClient()
 	if k8sErr != nil {
 		util.LogAndExit("Failed to create new kubernetes client, %s", k8sErr.Error())
 	}
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{ProtocolVersionConfigMapName: ProtocolVersionConfigMapName}}
 
 	var cm *corev1.ConfigMap
-	if configMaps, err := k8sClient.CoreV1().ConfigMaps("").List(metav1.ListOptions{}); err == nil {
-		for i, v := range configMaps.Items {
-			if v.Name == ProtocolVersionConfigMapName {
-				cm = &configMaps.Items[i]
-			}
+	configMaps, err := k8sClient.CoreV1().ConfigMaps("").List(
+		metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err == nil {
+		if len(configMaps.Items) > 1 {
+			return nil, errors.New("Multiple Cloudflow operators detected in the cluster. This is not supported. Exiting")
 		}
-	} else {
+		cm = &configMaps.Items[0]
+		if cm.GetName() == ProtocolVersionConfigMapName {
+			return cm, nil
+		}
+	}
+	return cm, err
+}
+
+// FindCloudflowNamespace tries to find the Cloudflow namespace set in the protocol version config map
+func FindCloudflowNamespace() (string, error) {
+	cm, err := GetProtocolVersionConfigMap()
+	if err != nil {
+		util.LogAndExit("Could not find the Cloudflow namespace. Kubernetes API returned an error: %s", err)
+	}
+
+	if cm == nil {
+		util.LogAndExit("Cannot find the '%s' ConfigMap and/or the Cloudflow namespace. Please make sure that the Cloudflow operator is installed", ProtocolVersionConfigMapName)
+	}
+	return cm.GetObjectMeta().GetNamespace(), err
+}
+
+// FailOnProtocolVersionMismatch fails and exits if the protocol version of kubectl-cloudflow does not match with the cloudflow operator protocol version.
+func FailOnProtocolVersionMismatch() {
```

Comment:
It looks like we are calling `GetProtocolVersionConfigMap` twice, once to get the namespace and once to verify the protocol version. Would it be possible to refactor this to only scan for the config map once?

Reply:
@agemooij `GetProtocolVersionConfigMap` is called twice in _different_ flow of control. If we want to avoid it we need to get the value and have it cached. Are you suggesting that ? I think that would be a micro-optimization as the call itself is not very expensive.

Reply:
I do not think Age is suggesting that we should cache the value, I think it's fine.